src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -411,6 +413,60 @@ def _generate_thread_status_messages(self, pool: ThreadPool, total_count: int):
             msgs.append("Processing Lines: " + ", ".join(lines) + ".")
         return msgs
 
+    def _determine_worker_count(self):
+        if not self._use_fork:
+            estimated_available_worker_count = get_available_max_worker_count()
+            n_process = min(self._worker_count, self._nlines, estimated_available_worker_count)
+            bulk_logger.info("Not using fork to create new process")
+            bulk_logger.info(
+                "The environment variable PF_WORKER_COUNT is not set or invalid. Calculate the worker count based on "
+                "the currently memory usage")
+            bulk_logger.info(
+                f"Calculated process count ({n_process}) by taking the minimum value among estimated process "
+                f"count ({estimated_available_worker_count}), the row count ({self._nlines}) and the default "
+                f"worker count ({self._worker_count}).")
+            return n_process
+        else:
+            n_process = min(self._worker_count, self._nlines)
+            bulk_logger.info("Using fork to create new process")
+            bulk_logger.info(
+                f"Calculated process count ({n_process}) by taking the minimum value among the the "
+                f"default value for worker_count ({self._worker_count}) and the row count ({self._nlines})")
+            return n_process
+
+    def load_worker_count_in_env(self):
```

Comment:
I think we can simply make this method return a valid positive int, or None otherwise.
We don't need to return `use_default_worker_count` to introduce extra complicity.

Maybe we can make it as a general utility method like:
![image](https://github.com/microsoft/promptflow/assets/113209/54a36aee-7b9d-4976-b59a-525e8cbda6d6)

We can put this utility method into a common file and write test cases for it to make it reliable.

Reply:
Added